Zend/zend_attributes.c

```diff
@@ -84,9 +70,52 @@ ZEND_API zend_attribute *zend_get_parameter_attribute_str(HashTable *attributes,
 	return get_attribute_str(attributes, str, len, offset + 1);
 }
 
-static void attribute_ptr_dtor(zval *v)
+static zend_always_inline void free_attribute(zend_attribute *attr, int persistent)
 {
-	zend_attribute_free((zend_attribute *) Z_PTR_P(v), 1);
+	uint32_t i;
+
+	zend_string_release(attr->name);
+	zend_string_release(attr->lcname);
+
+	for (i = 0; i < attr->argc; i++) {
+		zval_ptr_dtor(&attr->argv[i]);
+	}
+
+	pefree(attr, persistent);
+}
+
+static void attr_free(zval *v)
+{
+	free_attribute((zend_attribute *) Z_PTR_P(v), 0);
+}
+
+static void attr_pfree(zval *v)
+{
+	free_attribute((zend_attribute *) Z_PTR_P(v), 1);
+}
+
+ZEND_API zend_attribute *zend_add_attribute(HashTable **attributes, zend_bool persistent, uint32_t offset, zend_string *name, uint32_t argc)
+{
+	if (*attributes == NULL) {
+		*attributes = pemalloc(sizeof(HashTable), persistent);
+		zend_hash_init(*attributes, 8, NULL, persistent ? attr_pfree : attr_free, persistent);
+	}
+
+	zend_attribute *attr = pemalloc(ZEND_ATTRIBUTE_SIZE(argc), persistent);
+
+	if (persistent == (zend_bool) (GC_FLAGS(name) & IS_STR_PERSISTENT)) {
```

Comment:
This won't do what you're expecting. You need an explicit != 0 check for the flag.